contribute/design/1.x/cstor/pdb/20200113-cstor-poddisruptionbudget.md

```diff
@@ -0,0 +1,170 @@
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+  - [Proposal](#proposal)
+    - [User Stories](#user-stories)
+      - [HighAvailable Volumes Shouldn't be disturbed for Cluster Upgrade Activities](#highavailable-volumes-shouldnt-be-disturbed-for-cluster-upgrade-activities)
+    - [Proposed Implementation](#proposed-implementation)
+    - [Steps to perform user stories](#steps-to-perform-user-stories)
+    - [Low Level Design](#low-level-design)
+      - [Work Flow](#work-flow)
+      - [Sample Yaml](#sample-yaml)
+      - [Notes:](#notes)
+    - [TestCases](#testcases)
+  - [Drawbacks](#drawbacks)
+  - [Alternatives](#alternatives)
+  - [Infrastructure Needed](#infrastructure-needed)
+
+## Summary
+
+This proposal includes design details of Pod Disruption Budget(PDB) for cStor pool
+pods.
+- Not to allow more than quorum no.of HighAvailable(HA i.e volume provisioned
+  with >= 3 replicas) volume replicas pool pods to go down.
+
+## Motivation
+
+There are cases arising where OpenEBS users upgrade cluster or take out the
+multiple nodes for maintenance due to this HA volume is going to ReadOnly mode
+without informing users/admin[How? Multiple pool pod nodes are taken at time].
+
+### Goals
+
+- Create Pod Disruption Budget among the cStor pool pods where HA volume replicas exists.
+  Below are examples of valid voluntary disruptions
+  - Draining a node for repair or upgrade.
+  - Draining a node from a cluster to scale the cluster down.
+
+### Non-Goals
+
+- Invalid/Other valid voluntary disruptions are not supported via PDB.
+
+## Proposal
+
+### User Stories
+
+#### HighAvailable Volumes Shouldn't be disturbed for Cluster Upgrade Activities
+As an OpenEBS user, HA volumes shouldn't be disturbed for cluster upgrade or maintenance
+activities.
+
+### Proposed Implementation
+
+Kubernetes will send volume creation request to CSI driver when PVC is created.
+CSI driver will send a cStorVolume provision request via CVC CR to cStorVolumeConfig(CVC)
+controller. CVC controller will schedule the replicas among the pools after that
+CVC controller will create a PDB among the pool pods where the volume replicas
+are scheduled.
+
+Example:
+- Provisioned a CSPC with five pool specs that intern creates 5 cStor Pools.
+- Now, when 'n' no.of HA volumes are provisioned(with replica count as 3) then
+  volume replicas can be schedule in any available cStor pools.
+- After volume replicas are placed then cStorVolumeConfig controller will
+  create PDB among those cStor pools.
+- After successful creation of PDB corresponding cStorVolumeConfig will be
+  updated with PDB label.
+
+So when `n` no.of HA volumes are provisioned then sigma(nCr)[where n is no.of pools,
+r is no.of replicas] no.of PDB's will be created. For example if `n` pools are
+created then at max nC3 + nC4 + nC5 PDB's will be created(Where 3, 4 & 5 are
+replicas).
+
+Note: PDB will be created only for HA volumes(i.e only for the volumes greater
+than or equal to 3 replica count).
+
```

Comment:
Add here for deprovision case as well

Reply:
Already added in WorkFlow. Should I need to add over here also?